html/includes/authentication/sso.inc.php

```diff
@@ -0,0 +1,336 @@
+<?php
+/**
+ * sso.inc.php
+ *
+ * -Description-
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2017 Adam Bishop
+ * @author     Adam Bishop <adam@omega.org.uk>
+ */
+
+use LibreNMS\Config;
+use LibreNMS\Util\IP;
+use LibreNMS\Exceptions\AuthenticationException;
+use LibreNMS\Exceptions\InvalidIpException;
+
+/**
+ * Some functionality in this mechanism is inspired by confluence_http_authenticator (@chauth) and graylog-plugin-auth-sso (@Graylog)
+ *
+ */
+
+function init_auth()
+{
+    // No initialisation is required
+}
+
+function authenticate($username, $password)
+{
+    if (empty($username)) {
+        throw new AuthenticationException('sso.user_attr was not found or was empty');
+    }
+
+    // Build the user's details from attributes
+    $email = auth_sso_get_attr(Config::get('sso.email_attr'));
+    $realname = auth_sso_get_attr(Config::get('sso.realname_attr'));
+    $description = auth_sso_get_attr(Config::get('sso.descr_attr'));
+    $can_modify_passwd = 0;
+
+    $level = auth_sso_calculate_level();
+
+    // User has already been approved by the authenicator so if automatic user create/update is enabled, do it
+    if (Config::get('sso.create_users') && !user_exists($username)) {
```

Comment:
Is this block going to be called on every page and every graph load as we run `authenticate()` each time.

Reply:
I don't believe that's the case - I threw in a print statement and authenticate() only seems to be called on the first request. It gets short circuited if there's an active session.

If it did get called repeatedly (user with broken cookie settings) it's idempotent, and if a database cache is present, the impact is just an additional sql query per page load.

Reply:
Actually I've just remember we switched to using sessions once validated so you're right on this.